### PR TITLE
RpcApiProtocol and RpcServiceProtocol

### PR DESCRIPTION
**Pull request by altendky**
Draft for:
- [x] Cleanup `TODO`s
- [x] Reconsider `custom_get_connections()` name
- [x] On hold for some other adjustments to get more application of and checking of this hinting effort
  - [x] https://github.com/Chia-Network/chia-blockchain/pull/12172
  - [x] https://github.com/Chia-Network/chia-blockchain/pull/12204
  - [x] https://github.com/Chia-Network/chia-blockchain/pull/12153
  - Other granular follow-ups

### CODE REVIEW

**Review by xdustinface — COMMENTED** [review]
Shouldn't we also enforce those protocols now on all the related API and Services? Seems like its currently just partially enforced for some because of how hints are used. Might be a follow up PR or in here?

**Review by arvidn — COMMENTED** [review]
looks reasonable

File: chia/rpc/rpc_server.py [hunk 1]
```diff
@@ -37,10 +38,76 @@ class RpcEnvironment:
...
+
+
+class RpcServiceProtocol(Protocol):
+    _shut_down: bool
+
+    @property
+    def server(self) -> ChiaServer:
+        # a property so as to be read only which allows ChiaServer to satisfy
+        # Optional[ChiaServer]
+        pass
```
[the remark this answers is not in the document]
Comment (altendky) [reply chia/rpc/rpc_server.py:53]: Not beyond the explanation I gave that I thought one of our tools complained about `...`, but I'll try switching these.

[the remark this answers is not in the document]
Comment (altendky) [reply chia/rpc/rpc_server.py]: Maybe I was misremembering about trying to put the `...` on the same line as the definition and having black reformat it to the line after.

**Review by arvidn — DISMISSED** [review]

File: chia/full_node/full_node.py [hunk 2]
```diff
@@ -165,6 +174,40 @@ def __init__(
         self._transaction_queue_task = None
         self.simulator_transaction_callback = None
 
+    def get_connections(self, request_node_type: Optional[NodeType]) -> List[Dict[str, Any]]:
+        # TODO add peaks for peers
```
Comment (xdustinface) [inline chia/full_node/full_node.py:178]:
```suggestion
```

Looks like thats done?

File: chia/harvester/harvester.py [hunk 3]
```diff
@@ -28,16 +30,28 @@ class Harvester:
...
+        # This is a stop gap until the class usage is refactored such the values of
+        # integral attributes are known at creation of the instance.
+        if self._server is None:
+            raise RuntimeError("server not assigned")
+
+        return self._server
 
     def __init__(self, root_path: Path, config: Dict, constants: ConsensusConstants):
+        # TODO: hook this up to something?
+        self._shut_down = False
```
Comment (xdustinface) [inline chia/harvester/harvester.py:54]: Maybe just replace `Harvester._is_shutdown` with this in here?

File: chia/rpc/rpc_server.py [hunk 4]
```diff
@@ -226,8 +278,10 @@ async def open_connection(self, request: Dict[str, Any]) -> EndpointResult:
         on_connect = None
         if hasattr(self.rpc_api.service, "on_connect"):
             on_connect = self.rpc_api.service.on_connect
-        if getattr(self.rpc_api.service, "server", None) is None or not (
-            await self.rpc_api.service.server.start_client(target_node, on_connect)
+        if (
+            not hasattr(self.rpc_api.service, "server")
+            or self.rpc_api.service.server is None
+            or not await self.rpc_api.service.server.start_client(target_node, on_connect)
```
Comment (xdustinface) [inline chia/rpc/rpc_server.py:284]: Seems unrelated?

**Review by xdustinface — DISMISSED** [review]
Nice! Still see the three comments.

**Review by xdustinface — APPROVED** [review]

**Review by arvidn — APPROVED** [review]

**Review by emlowe — APPROVED** [review]
